source/opt/fix_storage_class.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "fix_storage_class.h"
+
+#include "source/opt/instruction.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status FixStorageClass::Process() {
+  bool modified = false;
+
+  get_module()->ForEachInst([this, &modified](Instruction* inst) {
+    if (inst->opcode() == SpvOpVariable) {
+      std::vector<Instruction*> uses;
+      get_def_use_mgr()->ForEachUser(
+          inst, [&uses](Instruction* use) { uses.push_back(use); });
+      for (Instruction* use : uses) {
+        modified |= PropagateStorageClass(
+            use, static_cast<SpvStorageClass>(inst->GetSingleWordInOperand(0)));
```

Comment:
Dunno if this can happen in an instruction.  What if there are 2 different storage classes in the instruction uses?

Reply:
This can happen for OpPhi instructions if the type are legitimately different storage types.  It will be wrong no matter what we do because the input operands have to match. We will end up picking one of them arbitrarily.  Since it cannot be fixed, the choice does not matter.  The same for an OpSelect.

All other instructions that we try to rewrite only have a single pointer operand, and the result type has to have the same storage class.